aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h

```diff
@@ -400,15 +459,76 @@ namespace impl {
     static_assert(std::is_same<guts::typelist::typelist<ParameterTypes...>, typename guts::infer_function_traits_t<KernelFunctor>::parameter_types>::value,
       "Parameter types mismatch");
 
-    static ReturnType call(OperatorKernel* functor, ParameterTypes... args) {
+    static ReturnType call(OperatorKernel* functor, DispatchKeySet, ParameterTypes... args) {
       KernelFunctor* functor_ = static_cast<KernelFunctor*>(functor);
+      // We're explicitly taking in a dispatchKeySet and not forwarding it to the registered kernel.
+      // This functor is used by all kernels that do not expect to take in a DispatchKeySet,
+      // i.e. pretty much all manually written kernels
       return (*functor_)(std::forward<ParameterTypes>(args)...);
     }
   };
 
+  // This specialization is for kernels with at least one argument.
+  // It's only difference is an extra static assert to check that the first argument isn't a DispatchKeySet
+  template<class KernelFunctor, class ReturnType, class FirstParam, class... ParameterTypes>
+  struct wrap_kernel_functor_unboxed_<KernelFunctor, ReturnType(FirstParam, ParameterTypes...)> final {
+    static_assert(!std::is_same<FirstParam, DispatchKeySet>::value,
+      "wrap_kernel_functor_unboxed expects the first argument of the schema NOT to be a DispatchKeySet. "
+      "If your kernel signature's first argument is a DispatchKeySet (and you plan to compute dispatch keys "
+      "explicitly in your kernel), you should register it with impl_withKeys().");
```

Comment:
I'm curious, what if instead of the `_withKeys` variants and these checks, we just type specialized on the first arg being DispatchKeySet directly? Maybe I'm forgetting something basic, but it seems like we have enough info from the signature to specialize what we need without explicit developer signaling via `impl`/`impl_withKeys` and their downstream counterparts

Reply:
That's exactly what I'm working on now :) Pushing a follow-up commit soon.

That should remove pretty much all of the duplication `library.h` and `KernelFunction.h` too- since the duplicated functions don't "do anything" different from the original, they just eventually forward to this different struct.